compiler/rustc_infer/src/infer/freshen.rs

```diff
@@ -140,79 +140,19 @@ impl<'a, 'tcx> TypeFolder<'tcx> for TypeFreshener<'a, 'tcx> {
         }
     }
 
+    #[inline]
     fn fold_ty(&mut self, t: Ty<'tcx>) -> Ty<'tcx> {
         if !t.needs_infer() && !t.has_erasable_regions() {
-            return t;
-        }
-
-        let tcx = self.infcx.tcx;
-
-        match *t.kind() {
-            ty::Infer(ty::TyVar(v)) => {
-                let opt_ty = self.infcx.inner.borrow_mut().type_variables().probe(v).known();
-                self.freshen_ty(opt_ty, ty::TyVar(v), ty::FreshTy)
-            }
+            t
+        } else {
+            match *t.kind() {
+                ty::Infer(v) => self.fold_infer_ty(v).unwrap_or(t),
 
-            ty::Infer(ty::IntVar(v)) => self.freshen_ty(
-                self.infcx
-                    .inner
-                    .borrow_mut()
-                    .int_unification_table()
-                    .probe_value(v)
-                    .map(|v| v.to_type(tcx)),
-                ty::IntVar(v),
-                ty::FreshIntTy,
-            ),
+                #[cfg(debug_assertions)]
```

Comment:
Personally think this should stay an "always" assertion

Reply:
agreed. To avoid any performance issues, this could call a `#[cold]` function with the `bug!` inside instead of having the formatting inside the main function

Reply:
I tried changing it back to an always assertion, and it had a noticeable perf impact, e.g. the instruction count for `wg-grammar` increased by 0.7%. I then tried the `#[cold]` function and it made a small improvement, but was still 0.5% worse.

So I will leave this as is, but I will add a comment about it.